contents/blog/missing-recordings.md

```diff
@@ -0,0 +1,58 @@
+---
+date: 2021-06-10
+title: The mystery of missing recordings at PostHog
+rootPage: /blog
+sidebar: Blog
+showTitle: true
+hideAnchor: true
+featuredImage: ../images/blog/
+featuredImageType: full
+author: ["rick-marron"]
+categories: ["Engineering"]
+---
+
+[Session Recording](product/session-recording) is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.
+
+This is the story of how we decreased the number of missing recordings from 15% to <1%.
+
+## Measuring the issue
+
+To start, we had to find a way to measure missing recordings. If we can't measure it, we can't know if we fixed it. But how can you measure something that's missing?
+
+We had a couple ideas, but they all had drawbacks. The two main ones we considered were:
+
+1. Measure how often we receive web analytics events outside of a recording. The idea is that a user should have a recording if they're sending up other events. The problem is that this doesn't account for cases when recordings are disabled, so the metric would always be < 100%.
+
+2. Measure how often we've received data for a recording, but it wasn't complete. For a recording to be shown to users, it needs to have enough data that we can draw the entire page - we call this a 'full snapshot'. This metric would measure how often we started a recording but didn't send up a full snapshot, so we couldn't render the page for the user. Theoretically, this should rarely happen.
+
+We decided to use the second metric because it would let us strive to 0% missing recordings. If you're curious, you can read the conversation that led to this decision in [this GitHub issue](https://github.com/PostHog/posthog/issues/5478).
+
+In the first run of the metric, it showed that we we're missing ~15% of recordings across [PostHog Cloud](/pricing) users.
+
+## Checking Sentry for clues
+
+Now that we had a metric, we started hunting for clues. What was causing the missing recordings?
+
+The first place we looked was Sentry. There were a couple of errors in the ingestion pipeline that seemed suspicious. The most notable was a "Max data upload size exceeded" error that was firing a few thousand times a day. This seemed like it could be an obvious cause of missing recordings. If some data was not being uploaded because it was too big, it would make sense that we could not get a 'full snapshot'.
+
+To figure out what was causing this error, we looked at data that had been uploaded that was just below our size threshold of 20mb. It looked like large data URLs of images were the cause.
+
+To test this out, we made a filter that would remove data URLs from any payload that exceeded a 5mb threshold, and it would replace the image with a simple SVG.
```

Comment:
Just looking at the PR, I think it would nice to add a little context over the SVG decision. There's a nice bit of back and forth and we can use the embedded images to add some visuals to the piece.

Reply:
Great idea. Pulled some of that discussion out into the post.